### PR TITLE
Optimize view_as_complex and view_as_real

### PR DESCRIPTION
**Pull request by peterbell10**
This avoids unnecessary memory allocations in `view_as_complex` and `view_as_real`. I construct the new tensor directly with the existing storage to avoid creating a new storage object and also use `DimVector`s to avoid allocating for the sizes and strides. Overall, this saves about 2 us of overhead from `torch.fft.fft` which currently has to call `view_as_real` and `view_as_complex` for every call.

I've used this simple benchmark to measure the overhead:
```python
In [1]: import torch
   ...: a = torch.rand(1, 2) 
   ...: ac = torch.view_as_complex(a) 
   ...: %timeit torch.view_as_real(ac)  
   ...: %timeit torch.view_as_complex(a) 
   ...: %timeit ac.real
```

Results before:
```
2.5 µs ± 62.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
2.22 µs ± 36 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
4.17 µs ± 8.76 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

and after:
```
1.83 µs ± 9.26 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
1.57 µs ± 7.17 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
3.47 µs ± 34.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

### CODE REVIEW

File: aten/src/ATen/native/ComplexHelper.h [hunk 1]
```diff
@@ -4,12 +4,25 @@
...
+  return new_tensor;
+}
+
+inline DimVector computeStrideForViewAsReal(IntArrayRef oldstride) {
+  DimVector res(oldstride.size() + 1);
+  for(size_t i = 0; i < oldstride.size(); i++) {
+    res[i] = oldstride[i] * 2;
   }
-  res.emplace_back(1);
+  res.back() = 1;
```
Comment (anjali411) [inline aten/src/ATen/native/ComplexHelper.h:25]: just curious if there's a reason why `back()` is more preferable thank `emplace_back()`

**Review by anjali411 — APPROVED** [review]
LGTM thank you!

File: aten/src/ATen/native/ComplexHelper.h (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/ComplexHelper.h:25]: @peterbell10 pre-allocated the size of `res` to avoid a (possible) reallocation

Comment (peterbell10) [reply aten/src/ATen/native/ComplexHelper.h:25]: Yeah, `emplace_back` inserts a new element while `back` is just a shorthand for indexing the last element.

**Comment by anjali411** [comment]
@peterbell10 I see some internal test failures which are probably unrelated. can you rebase this PR?

**Pull request closed without merging**